### DISCUSSION TITLE
Test fails after migrating from v6 to v7. Date type changed from string to object.

### DISCUSSION THREAD

**Discussion by leochoo** [react-hook-form/react-hook-form#8217] [Q&A] [answered]
@Moshyfawn @bluebill1049 This is a follow-up to the issue I opened before. #8144 
### Repo
v6: https://codesandbox.io/s/reproduce-v6-j5qnfi?file=/src/use-form-date.test.tsx
v7: https://codesandbox.io/s/reproduce-v7-1j4342?file=/src/use-form-date.test.tsx:268-303

# Issue: Test works on v6 but not on v7.
Above CSB I shared for #8144 contains the file called `use-form-date.test.tsx`.

These are the results.
I have absolutely no idea why the Date was returned as `string` in v6 and not it is kept as `Dayjs` object in v7.
I was wondering if #4988 is related (hence I commented once).

I would greatly appreciate some help. Thank you!

v6
![CleanShot 2022-04-06 at 18 11 13](https://user-images.githubusercontent.com/26427048/163908113-7104f8cb-0b57-4237-bd9f-1e2a1c303e03.png)

v7
![CleanShot 2022-04-06 at 18 11 28](https://user-images.githubusercontent.com/26427048/163908119-dd3c7446-cd5c-4c92-909b-6ace8cfbb9f9.png)

# Important
[Due to a CSB limitation](https://github.com/codesandbox/codesandbox-client/issues/6599), you CANNOT run this test directly on CSB.

Please do `File > Export to ZIP` and run it on your local machine.

**Comment by bluebill1049** [comment]
it really depends on your input and how it's been wired up, the received value seems to be correct for native input unless you have some sort of parsing that will return `Date` type value.

**Reply by Moshyfawn** [reply]
IDK, I've trying a simple CSB from scratch and V6 does seem to parse the DayJS class into a date `string` when `setValue("fieldName", dayjs("2021-03-21"))`.

I haven't worked with V6 all that much, so I'm not sure what's going on here

**Reply by leochoo** [reply]
There is no parsing or transforming done on my side. RHF version is the only difference between the two.

So do you think it is safe to change my test so that it expects an `Object` instead of a `string`?

**Reply by bluebill1049** [reply]
i will go with string any day for input.

**Comment by leapful** [comment, marked as answer]
@leochoo , @bluebill1049 , @Moshyfawn : 

The default handling of "react-hook-form" will set value to "ref.value" (HTML field element) when calling "setValue". For a complex object like dayjs object, method "toString" will be called to stringify the value.

With version 6, "getValues" will return "ref.value" by default therefore it's in date string format.

With version 7, "getValues" will return the dayjs object from "formState", not from "ref.value" anymore, that why your unit test is failed.

I would suggest to use only string or number for handling date value in field element instead of an object like dayjs object.

NOTE: Jest uses "pretty-format" to stringify the "received value" of "expect" and that library uses "toJSON" to transform dayjs object for comparing, that's why it show date string in ISO format for the error message.

**Reply by bluebill1049** [reply]
Someone knows this library better than I do. 🎩 off

**Reply by Moshyfawn** [reply]
Nice bit of research, thx! I was suspecting it might be something to do with getting the values from the actual input, but I'm too lazy to look into the source of code of an old lib version lol
